lib/grandpa/finalization_test.go

```diff
@@ -0,0 +1,260 @@
+				builder := func() (engine ephemeralService, voting ephemeralService) {
+					mockEngine := NewMockephemeralService(ctrl)
+					mockEngine.EXPECT().Start().DoAndReturn(func() error {
+						<-time.NewTimer(3 * time.Second).C
+						return nil
+					})
+					mockEngine.EXPECT().Stop().Return(nil)
+
+					votingStopCh := make(chan struct{})
+					mockVoting := NewMockephemeralService(ctrl)
+					mockVoting.EXPECT().Start().DoAndReturn(func() error {
+						<-votingStopCh
+						return nil
+					})
+					mockVoting.EXPECT().Stop().DoAndReturn(func() error {
+						close(votingStopCh)
+						return nil
+					})
+
+					return mockEngine, mockVoting
+				}
+
+				return &finalizationHandler{
+					newServices:    builder,
+					timeoutStop:    2 * time.Second,
+					observableErrs: make(chan error),
+					stopCh:         make(chan struct{}),
+					handlerDone:    make(chan struct{}),
+				}
+			},
+		},
+
+		"voting_round_fails_should_stop_engine_service": {
+			wantErr: errors.New("mocked voting round fails"),
+			createFinalizationHandler: func(ctrl *gomock.Controller) *finalizationHandler {
+				builder := func() (engine ephemeralService, voting ephemeralService) {
+					failTime := 2 * time.Second
+
+					mockVoting := NewMockephemeralService(ctrl)
+					mockVoting.EXPECT().Start().DoAndReturn(func() error {
+						timeToFail := time.NewTimer(failTime)
+						<-timeToFail.C
+						return errors.New("mocked voting round fails")
+					})
+					mockVoting.EXPECT().Stop().Return(nil)
+
+					// once the voting round fails the finalisation handler
+					// should be awere of the error and call the stop method from
+					// the engine which will release the start method from engine service
+					engineStopCh := make(chan struct{})
+					mockEngine := NewMockephemeralService(ctrl)
+					mockEngine.EXPECT().Start().DoAndReturn(func() error {
+						select {
+						case <-time.After(failTime + time.Second):
+							return errors.New("timeout waiting engineStopCh")
+						case <-engineStopCh:
+							return nil
+						}
+					})
+					mockEngine.EXPECT().Stop().DoAndReturn(func() error {
+						close(engineStopCh)
+						return nil
+					})
+
+					return mockEngine, mockVoting
+				}
+
+				return &finalizationHandler{
+					newServices:    builder,
+					timeoutStop:    2 * time.Second,
+					observableErrs: make(chan error),
+					stopCh:         make(chan struct{}),
+					handlerDone:    make(chan struct{}),
+				}
+			},
+		},
+
+		"engine_fails_should_stop_voting_round_service": {
+			wantErr: errors.New("mocked finalisation engine fails"),
+			createFinalizationHandler: func(ctrl *gomock.Controller) *finalizationHandler {
+				builder := func() (engine ephemeralService, voting ephemeralService) {
+					failTime := 2 * time.Second
+
+					mockEngine := NewMockephemeralService(ctrl)
+					mockEngine.EXPECT().Start().DoAndReturn(func() error {
+						timeToFail := time.NewTimer(failTime)
+						<-timeToFail.C
```

Comment:
nit / applies to other `timeToFail` in this file

```suggestion
						time.Sleep(failTime)
```